README.md
# grunt-cognos-ext-upload

[![CircleCI](https://circleci.com/gh/CognosExt/grunt-cognos-ext-upload.svg?style=svg)](https://circleci.com/gh/CognosExt/grunt-cognos-ext-upload) [![Greenkeeper badge](https://badges.greenkeeper.io/CognosExt/grunt-cognos-ext-upload.svg)](https://greenkeeper.io/)

Use [Grunt](http://gruntjs.com/) to upload your [IBM Cognos Analytics](https://www.ibm.com/products/cognos-analytics) custom [extensions](https://www.ibm.com/support/knowledgecenter/SSEP7J_11.0.0/com.ibm.swg.ba.cognos.ag_manage.doc/c_ag_manage_extensions.html) directly from your local machine into your cognos installation without opening your browser.

## Getting Started

This plugin requires Grunt `>=0.4.0`

If you haven't used [Grunt](http://gruntjs.com/) before, be sure to check out the [Getting Started](http://gruntjs.com/getting-started) guide, as it explains how to create a [Gruntfile](http://gruntjs.com/sample-gruntfile) as well as install and use Grunt plugins. Once you're familiar with that process, you may install this plugin with this command:

```shell
npm install grunt-cognos-ext-upload --save-dev
```

Once the plugin has been installed, it may be enabled inside your Gruntfile with this line of JavaScript:

```javascript
grunt.loadNpmTasks("grunt-documentation");
```

# Usage

<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### Table of Contents

-   [gruntUpload](#gruntupload)
    -   [Parameters](#parameters)
    -   [Examples](#examples)

## gruntUpload

Grunt Cognos Extension Upload

In your project's Gruntfile, add a section named `cognos_ext_upload` to the
data object passed into `grunt.initConfig()`.

### Parameters

-   `grunt`  
-   `options` **[Object](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object)** 
    -   `options.name` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Name of the extension as found in the specs.json (TODO: read the name from the specs.json)
    -   `options.user` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Cognos Username with enough priviliges to upload (new) extensions
    -   `options.password` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Password of the user
    -   `options.namespace` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** The Cognos namespace id, if not the default
    -   `options.url` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** URL of the homepage of your Cognos 11 installation (eg. <https://localhost/ibmcognos> )
    -   `options.type` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** type of upload. Default is 'extensions', for themes use 'themes'.
    -   `options.zipfile` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** name of the zipfile to upload. Defaults to dist/extension.zip
    -   `options.timeout` **[number](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Number)** Timeout of connection in milliseconds, defaults to 60000
    -   `options.ignoreInvalidCertificates` **[boolean](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Boolean)** Ignore invalid ssl certificates. Default is false.
    -   `options.debug` **[string](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String)** Creates more output

### Examples

```javascript
grunt.initConfig({
  documentation: {
      default: {
          options: {
              name: 'My_Theme',
              user: "admin",
              password: "secret",
              namespace: "MyNameSpace",
              url: "https://localhost/ibmcognos",
              type: "themes",
              zipfile: "dist/mytheme.zip",
              timeout: 30000,
              ignoreInvalidCertificates: true,
              debug: false
          }
      },
  }
});
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.7.1](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.7.0...v1.7.1) (2020-06-05)

## [1.7.0](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.6.2...v1.7.0) (2019-09-05)


### Features

* typo ([ad9b87d](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/ad9b87d))

### [1.6.2](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.6.1...v1.6.2) (2019-09-05)


### Bug Fixes

* throw a proper grunt error when upload fails ([476a7ee](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/476a7ee))

### [1.6.1](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.6.0...v1.6.1) (2019-08-15)

<a name="1.6.0"></a>
# [1.6.0](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.5.2...v1.6.0) (2019-07-11)


### Features

* Added new option for timeout and changed name of option to ignore invalid ssl certificates ([10f0724](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/10f0724))



<a name="1.5.2"></a>
## [1.5.2](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.5.1...v1.5.2) (2019-07-10)


### Bug Fixes

* Fix a dependency ([18580da](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/18580da))



<a name="1.5.1"></a>
## [1.5.1](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.5.0...v1.5.1) (2019-07-10)


### Bug Fixes

* Update dependency with bug ([fba186b](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/fba186b))



<a name="1.5.0"></a>
# [1.5.0](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.4.1...v1.5.0) (2019-07-10)


### Features

* Added options to ignore invalid ssl certificates ([88308f8](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/88308f8))



<a name="1.4.1"></a>
## [1.4.1](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.4.0...v1.4.1) (2018-12-12)


### Bug Fixes

* Never called done() so no other task would ever execute after this one ([d1be8ad](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/d1be8ad))



<a name="1.4.0"></a>
# [1.4.0](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.2.2...v1.4.0) (2018-10-26)


### Features

* Namespace support ([3b52133](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/3b52133))



<a name="1.2.2"></a>

## [1.2.2](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.2.1...v1.2.2) (2018-10-26)

<a name="1.2.1"></a>

## [1.2.1](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.2.0...v1.2.1) (2018-10-26)

<a name="1.2.0"></a>

# [1.2.0](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.1.4...v1.2.0) (2018-07-03)

### Features

- Added the option to also upload themes during development ([7bcfae7](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/7bcfae7))

<a name="1.1.4"></a>

## [1.1.4](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.1.3...v1.1.4) (2018-06-20)

<a name="1.1.3"></a>

## [1.1.3](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.1.2...v1.1.3) (2018-06-20)

<a name="1.1.2"></a>

## [1.1.2](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.1.1...v1.1.2) (2017-10-27)

<a name="1.1.1"></a>

## [1.1.1](https://github.com/CognosExt/grunt-cognos-ext-upload/compare/v1.1.0...v1.1.1) (2017-10-26)

<a name="1.1.0"></a>

# 1.1.0 (2017-10-26)

### Features

- Initial Commit ([6ccd566](https://github.com/CognosExt/grunt-cognos-ext-upload/commit/6ccd566))


test/test.js
'use strict';

exports.cognosextupload = {
  run: function(test) {
    // not sure what to test yet, lets say we are happy
    test.ok(true, true);
    test.done();
  }
};


tasks/cognos-ext-upload.js
'use strict';

/**
 * Grunt Cognos Extension Upload
 *
 * In your project's Gruntfile, add a section named `cognos_ext_upload` to the
 * data object passed into `grunt.initConfig()`.
 *
 * @param {Object} options
 * @param {string} options.name Name of the extension as found in the specs.json (TODO: read the name from the specs.json)
 * @param {string} options.user Cognos Username with enough priviliges to upload (new) extensions
 * @param {string} options.password Password of the user
 * @param {string} options.namespace The Cognos namespace id, if not the default
 * @param {string} options.url URL of the homepage of your Cognos 11 installation (eg. https://localhost/ibmcognos )
 * @param {string} options.type type of upload. Default is 'extensions', for themes use 'themes'.
 * @param {string} options.zipfile name of the zipfile to upload. Defaults to dist/extension.zip
 * @param {number} options.timeout Timeout of connection in milliseconds, defaults to 60000
 * @param {boolean} options.ignoreInvalidCertificates Ignore invalid ssl certificates. Default is false.
 * @param {string} options.debug Creates more output
 * @example
 * grunt.initConfig({
 *   documentation: {
 *       default: {
 *           options: {
 *               name: 'My_Theme',
 *               user: "admin",
 *               password: "secret",
 *               namespace: "MyNameSpace",
 *               url: "https://localhost/ibmcognos",
 *               type: "themes",
 *               zipfile: "dist/mytheme.zip",
 *               timeout: 30000,
 *               ignoreInvalidCertificates: true,
 *               debug: false
 *           }
 *       },
 *   }
 * });
 */
function gruntUpload(grunt) {
  grunt.registerMultiTask(
    'cognos_ext_upload',
    'Use Grunt to upload a cognos extensions zip file.',
    function() {
      var options = this.options({
        name: '',
        user: '',
        password: '',
        namespace: '',
        url: '',
        type: 'extensions',
        zipfile: 'dist/extension.zip',
        timeout: 60000,
        ignoreInvalidCertificates: false,
        debug: false
      });

      grunt.log.writeln('Going to upload ' + options.name);
      var done = this.async();

      const name = options.name;
      const zipfile = options.zipfile;
      const url = options.url;
      const debug = options.debug;
      const user = options.user;
      const password = options.password;
      const namespace = options.namespace;
      const exttype = options.type;
      const timeout = options.timeout;
      const ignoreInvalidCertificates = options.ignoreInvalidCertificates;

      const jcognos = require('jcognos');
      var getCognos = jcognos.getCognos;

      var result = getCognos(url, debug, timeout, ignoreInvalidCertificates)
        .then(function(lcognos) {
          lcognos.login(user, password, namespace).then(function() {
            grunt.log.writeln('going to upload ' + zipfile);
            if (options.debug) {
              grunt.log.writeln('going to upload ' + name);
              grunt.log.writeln('going to type ' + exttype);
            }
            lcognos
              .uploadExtension(zipfile, name, exttype)
              .then(function() {
                console.log('Uploaded Extension');
                done();
              })
              .catch(function(err) {
                console.log('Error uploading', err);
              });
          });
        })
        .catch(function(error) {
          console.log(error);
          grunt.log.writeln('error: ' + error);
          var util = require('util');
          done(util.isError(error) ? error : new Error(error));
        });
    }
  );
}

module.exports = gruntUpload;
